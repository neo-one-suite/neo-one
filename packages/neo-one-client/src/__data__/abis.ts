import { ABI, ABIEvent, ABIFunction, ABIParameter, ABIReturn } from '../types';

export const returns = {
  Signature: { type: 'Signature' } as ABIReturn,
  Boolean: { type: 'Boolean' } as ABIReturn,
  Hash160: { type: 'Hash160' } as ABIReturn,
  Hash256: { type: 'Hash256' } as ABIReturn,
  ByteArray: { type: 'ByteArray' } as ABIReturn,
  PublicKey: { type: 'PublicKey' } as ABIReturn,
  String: { type: 'String' } as ABIReturn,
  InteropInterface: { type: 'InteropInterface' } as ABIReturn,
  Void: { type: 'Void' } as ABIReturn,
  Integer: { type: 'Integer', decimals: 7 } as ABIReturn,
  Array: { type: 'Array', value: { type: 'String' } } as ABIReturn,
};

export const parameters = {
  Signature: { name: 'signatureName', type: 'Signature' } as ABIParameter,
  Boolean: { name: 'booleanName', type: 'Boolean' } as ABIParameter,
  Integer: { name: 'intName', type: 'Integer', decimals: 7 } as ABIParameter,
  Hash160: { name: 'hash160Name', type: 'Hash160' } as ABIParameter,
  Hash256: { name: 'hash256Name', type: 'Hash256' } as ABIParameter,
  ByteArray: { name: 'byteName', type: 'ByteArray' } as ABIParameter,
  PublicKey: { name: 'pubkeyName', type: 'PublicKey' } as ABIParameter,
  String: { name: 'stringName', type: 'String' } as ABIParameter,
  Array: {
    name: 'arrayName',
    type: 'Array',
    value: {
      name: 'stringName',
      type: 'String',
    },
  } as ABIParameter,

  InteropInterface: {
    name: 'interopName',
    type: 'InteropInterface',
  } as ABIParameter,
  Void: { name: 'voidName', type: 'Void' } as ABIParameter,
};

export const abiFunction = (
  constant = false,
  params: ABIParameter[] = [parameters.String],
  verify = false,
  name = 'funcName',
  returnType: ABIReturn = returns.Void,
): ABIFunction => ({
  name,
  constant,
  parameters: params,
  verify,
  returnType,
});

export const abiEvent = (name = 'eventName', params: ABIParameter[] = [parameters.String]): ABIEvent => ({
  name,
  parameters: params,
});

export const abi = (functions: ABIFunction[] = [abiFunction()], events: ABIEvent[] = []): ABI => ({
  functions,
  events,
});
