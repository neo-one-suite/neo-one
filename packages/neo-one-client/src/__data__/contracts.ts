import BigNumber from 'bignumber.js';
import { bigNumberToBN } from '../utils';

export const contracts = {
  Signature: {
    type: 'Signature',
    value: '02028a99826ed',
  },
  Boolean: { type: 'Boolean', value: true },
  Integer: {
    type: 'Integer',
    value: bigNumberToBN(new BigNumber('10'), 10),
  },
  Hash160: {
    type: 'Hash160',
    value: '0x3775292229eccdf904f16fff8e83e7cffdc0f0ce',
  },
  Hash256: {
    type: 'Hash256',
    value: '0x798fb9e4f3437e4c9ab83c24f950f0fce98e1d1aaac4fe3f54a66c5d9def89ca',
  },
  ByteArray: {
    type: 'ByteArray',
    value: '02028a99826ed',
  },
  PublicKey: {
    type: 'PublicKey',
    value: '02028a99826edc0c97d18e22b6932373d908d323aa7f92656a77ec26e8861699ef',
  },
  String: {
    type: 'String',
    value: 'stringVal',
  },
  Array: {
    type: 'Array',
    value: [{ type: 'String', value: 'stringVal' }],
  },
  InteropInterface: { type: 'InteropInterface' },
  Void: { type: 'Void' },
};
