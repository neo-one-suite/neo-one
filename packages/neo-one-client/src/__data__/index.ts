export * from './keys';
export * from './transactions';
