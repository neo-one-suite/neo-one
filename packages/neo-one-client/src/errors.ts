import { CustomError } from '@neo-one/utils';
import BigNumber from 'bignumber.js';
import { ContractParameter, ContractParameterType } from './types';

export class InvalidContractParameterError extends CustomError {
  public readonly parameter: ContractParameter;
  public readonly expected: ReadonlyArray<ContractParameterType>;
  public readonly code: string;

  public constructor(parameter: ContractParameter, expected: ReadonlyArray<ContractParameterType>) {
    super(`Expected one of ${JSON.stringify(expected)} ` + `ContractParameterTypes, found ${parameter.type}`);

    this.parameter = parameter;
    this.expected = expected;
    this.code = 'INVALID_CONTRACT_PARAMETER';
  }
}

export class InvalidArgumentError extends CustomError {
  public readonly code: string;

  public constructor(message: string) {
    super(message);
    this.code = 'INVALID_ARGUMENT';
  }
}

export class InvalidNamedArgumentError extends InvalidArgumentError {
  public readonly code: string;

  public constructor(name: string, argument: {}) {
    super(`Invalid argument for ${name}: ${String(argument)}`);
    this.code = 'INVALID_NAMED_ARGUMENT';
  }
}

export class InvocationCallError extends CustomError {
  public readonly code: string;

  public constructor(message: string) {
    super(message);
    this.code = 'INVOCATION_CALL';
  }
}

export class InvalidEventError extends CustomError {
  public readonly code: string;

  public constructor(message: string) {
    super(message);
    this.code = 'INVALID_EVENT';
  }
}

export class InsufficientFundsError extends CustomError {
  public readonly code: string;

  public constructor(total: BigNumber, expected: BigNumber) {
    super(`Found ${total.toString()} funds, required: ${expected.toString()}.`);
    this.code = 'INSUFFICIENT_FUNDS';
  }
}

export class FundsInUseError extends CustomError {
  public readonly code: string;

  public constructor(total: BigNumber, expected: BigNumber, numInputs: number) {
    super(
      `Found ${total.toString()} funds, required: ${expected.toString()}; You have ${numInputs} input(s) on the current block, try transfer again on the next`,
    );
    this.code = 'FUNDS_IN_USE';
  }
}

export class NothingToClaimError extends CustomError {
  public readonly code: string;

  public constructor() {
    super('Nothing to claim.');
    this.code = 'NOTHING_TO_CLAIM';
  }
}

export class NothingToIssueError extends CustomError {
  public readonly code: string;

  public constructor() {
    super('Nothing to issue.');
    this.code = 'NOTHING_TO_ISSUE';
  }
}

export class NothingToTransferError extends CustomError {
  public readonly code: string;

  public constructor() {
    super('Nothing to transfer.');
    this.code = 'NOTHING_TO_TRANSFER';
  }
}

export class NoAccountError extends CustomError {
  public readonly code: string;

  public constructor() {
    super('No account exists.');
    this.code = 'NO_ACCOUNT';
  }
}

export class NoContractDeployedError extends CustomError {
  public readonly code: string;

  public constructor(networkType: string) {
    super(`Contract has not been deployed to network ${networkType}`);
    this.code = 'NO_CONTRACT_DEPLOYED';
  }
}

export class InvalidTransactionError extends CustomError {
  public readonly code: string;

  public constructor(message: string) {
    super(message);
    this.code = 'INVALID_TRANSACTION';
  }
}

export class InvokeError extends CustomError {
  public readonly code: string;

  public constructor(message: string) {
    super(message);
    this.code = 'INVOKE';
  }
}

export class UnknownBlockError extends CustomError {
  public readonly code: string;

  public constructor() {
    super('Unknown block');
    this.code = 'UNKNOWN_BLOCK';
  }
}

export class RelayTransactionError extends CustomError {
  public readonly code: string;

  public constructor(message: string) {
    super(message);
    this.code = 'RELAY_TRANSACTION';
  }
}

export class UnknownNetworkError extends CustomError {
  public readonly code: string;

  public constructor(name: string) {
    super(`Unknown network ${name}`);
    this.code = 'UNKNOWN_NETWORK';
  }
}

export class UnknownAccountError extends CustomError {
  public readonly code: string;

  public constructor(address: string) {
    super(`Unknown account ${address}`);
    this.code = 'UNKNOWN_ACCOUNT';
  }
}

export class LockedAccountError extends CustomError {
  public readonly code: string;

  public constructor(address: string) {
    super(`Account ${address} is locked`);
    this.code = 'LOCKED_ACCOUNT';
  }
}

export class PasswordRequiredError extends CustomError {
  public readonly code: string;

  public constructor() {
    super('A password is required when creating accounts on the MainNet.');
    this.code = 'PASSWORD_REQUIRED';
  }
}
