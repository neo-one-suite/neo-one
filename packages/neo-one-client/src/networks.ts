export const MAIN = 'main';
export const MAIN_URL = 'https://neotracker.io/rpc';
export const TEST = 'test';
export const TEST_URL = 'https://testnet.neotracker.io/rpc';
