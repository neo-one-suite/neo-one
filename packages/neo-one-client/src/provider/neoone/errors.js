/* @flow */
import { CustomError } from '@neo-one/utils';

export type JSONRPCErrorResponse = {|
  code: number,
  message: string,
  data?: $FlowFixMe,
|};

export class JSONRPCError extends CustomError {
  responseError: JSONRPCErrorResponse;
  code: string;

  constructor(responseError: JSONRPCErrorResponse) {
    super(responseError.message);
    this.responseError = responseError;
    this.code = 'JSON_RPC';
  }
}

export class InvalidRPCResponseError extends CustomError {
  code: string;
  constructor() {
    super('Did not receive valid rpc response');
    this.code = 'INVALID_RPC_RESPONSE';
  }
}

export class HTTPError extends CustomError {
  status: number;
  text: ?string;
  code: string;

  constructor(status: number, text: ?string) {
    let message = `HTTP Error ${status}`;
    if (text != null) {
      message = `${message}: ${text}`;
    }
    super(message);
    this.status = status;
    this.text = text;
    this.code = 'HTTP';
  }
}

export class MissingTransactionDataError extends CustomError {
  code: string = 'MISSING_TRANSACTION_DATA';

  constructor(hash: string) {
    super(`Missing transaction data for transaction ${hash}`);
  }
}
