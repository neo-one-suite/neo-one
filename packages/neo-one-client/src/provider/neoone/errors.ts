import { CustomError } from '@neo-one/utils';

export interface JSONRPCErrorResponse {
  readonly code: number;
  readonly message: string;
  // tslint:disable-next-line no-any
  readonly data?: any;
}

export class JSONRPCError extends CustomError {
  public readonly responseError: JSONRPCErrorResponse;
  public readonly code: string;

  public constructor(responseError: JSONRPCErrorResponse) {
    super(responseError.message);
    this.responseError = responseError;
    this.code = 'JSON_RPC';
  }
}

export class InvalidRPCResponseError extends CustomError {
  public readonly code: string;

  public constructor() {
    super('Did not receive valid rpc response');
    this.code = 'INVALID_RPC_RESPONSE';
  }
}

export class HTTPError extends CustomError {
  public readonly status: number;
  public readonly text: string | undefined;
  public readonly code: string;

  public constructor(status: number, text?: string) {
    super(text === undefined ? `HTTP Error ${status}` : `HTTP Error ${status}: ${text}`);
    this.status = status;
    this.text = text;
    this.code = 'HTTP';
  }
}

export class MissingTransactionDataError extends CustomError {
  public readonly code: string = 'MISSING_TRANSACTION_DATA';

  public constructor(hash: string) {
    super(`Missing transaction data for transaction ${hash}`);
  }
}
