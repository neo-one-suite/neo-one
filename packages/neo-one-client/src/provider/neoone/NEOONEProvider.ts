import { Monitor } from '@neo-one/monitor';
import BigNumber from 'bignumber.js';
import { BehaviorSubject, Observable } from 'rxjs';
import { UnknownNetworkError } from '../../errors';
import * as networkConfigs from '../../networks';
import {
  AddressString,
  GetOptions,
  Hash256String,
  Input,
  NetworkSettings,
  NetworkType,
  RawInvocationData,
  RawInvocationResult,
  Transaction,
  TransactionReceipt,
  UnspentOutput,
} from '../../types';
import { NEOONEDataProvider } from './NEOONEDataProvider';

export interface NEOONEProviderOptions {
  readonly network: NetworkType;
  readonly rpcURL: string;
}

export class NEOONEProvider {
  public readonly networks$: Observable<ReadonlyArray<NetworkType>>;
  private readonly networksInternal$: BehaviorSubject<ReadonlyArray<NetworkType>>;
  private readonly mutableProviders: { [K in string]?: NEOONEDataProvider };

  public constructor(
    input: {
      readonly mainRPCURL?: string;
      readonly testRPCURL?: string;
      readonly options?: ReadonlyArray<NEOONEProviderOptions>;
    } = {},
  ) {
    const { mainRPCURL = networkConfigs.MAIN_URL, testRPCURL = networkConfigs.TEST_URL, options = [] } = input;
    this.networksInternal$ = new BehaviorSubject([] as ReadonlyArray<NetworkType>);
    this.networks$ = this.networksInternal$;
    this.mutableProviders = {};

    // tslint:disable-next-line no-let
    let hasMain = false;
    // tslint:disable-next-line no-let
    let hasTest = false;
    // tslint:disable-next-line no-let
    let networks = options.map(({ network, rpcURL }) => {
      if (network === networkConfigs.MAIN) {
        hasMain = true;
      }

      if (network === networkConfigs.TEST) {
        hasTest = true;
      }

      this.mutableProviders[network] = new NEOONEDataProvider({
        network,
        rpcURL,
      });

      return network;
    });

    if (!hasMain) {
      this.mutableProviders.main = new NEOONEDataProvider({
        network: networkConfigs.MAIN,
        rpcURL: mainRPCURL,
      });

      networks = networks.concat([networkConfigs.MAIN]);
    }

    if (!hasTest) {
      this.mutableProviders.test = new NEOONEDataProvider({
        network: networkConfigs.TEST,
        rpcURL: testRPCURL,
      });

      networks = networks.concat([networkConfigs.TEST]);
    }

    this.networksInternal$.next(networks);
  }

  public getNetworks(): ReadonlyArray<NetworkType> {
    return this.networksInternal$.getValue();
  }

  public addNetwork({ network, rpcURL }: { readonly network: NetworkType; readonly rpcURL: string }): void {
    this.mutableProviders[network] = new NEOONEDataProvider({ network, rpcURL });
    const networks = this.networksInternal$.value.filter((net) => network !== net).concat([network]);
    this.networksInternal$.next(networks);
  }

  public async getUnclaimed(
    network: NetworkType,
    address: AddressString,
    monitor?: Monitor,
  ): Promise<{ readonly unclaimed: ReadonlyArray<Input>; readonly amount: BigNumber }> {
    return this.getProvider(network).getUnclaimed(address, monitor);
  }

  public async getUnspentOutputs(
    network: NetworkType,
    address: AddressString,
    monitor?: Monitor,
  ): Promise<ReadonlyArray<UnspentOutput>> {
    return this.getProvider(network).getUnspentOutputs(address, monitor);
  }

  public async relayTransaction(network: NetworkType, transaction: string, monitor?: Monitor): Promise<Transaction> {
    return this.getProvider(network).relayTransaction(transaction, monitor);
  }

  public async getTransactionReceipt(
    network: NetworkType,
    hash: Hash256String,
    options?: GetOptions,
  ): Promise<TransactionReceipt> {
    return this.getProvider(network).getTransactionReceipt(hash, options);
  }

  public async getInvocationData(
    network: NetworkType,
    hash: Hash256String,
    monitor?: Monitor,
  ): Promise<RawInvocationData> {
    return this.getProvider(network).getInvocationData(hash, monitor);
  }

  public async testInvoke(network: NetworkType, transaction: string, monitor?: Monitor): Promise<RawInvocationResult> {
    return this.getProvider(network).testInvoke(transaction, monitor);
  }

  public async getNetworkSettings(network: NetworkType, monitor?: Monitor): Promise<NetworkSettings> {
    return this.getProvider(network).getNetworkSettings(monitor);
  }

  public async getBlockCount(network: NetworkType, monitor?: Monitor): Promise<number> {
    return this.getProvider(network).getBlockCount(monitor);
  }

  public read(network: NetworkType): NEOONEDataProvider {
    return this.getProvider(network);
  }

  private getProvider(network: NetworkType): NEOONEDataProvider {
    const provider = this.mutableProviders[network];
    if (provider === undefined) {
      throw new UnknownNetworkError(network);
    }

    return provider;
  }
}
