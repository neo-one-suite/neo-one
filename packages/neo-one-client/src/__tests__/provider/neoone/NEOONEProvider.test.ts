import { keys, transactions } from '../../../__data__';
import { UnknownNetworkError } from '../../../errors';
import * as networkConfigs from '../../../networks';
import { NEOONEProvider } from '../../../provider/neoone/NEOONEProvider';

describe('NEOONEProvider', () => {
  const network = 'net';
  const expected = '0';
  const mainRPCURL = 'foo';
  const testRPCURL = 'bar';
  const options = [{ network, rpcURL: 'rpc' }];

  let provider = new NEOONEProvider({ mainRPCURL, testRPCURL, options });
  beforeEach(() => {
    provider = new NEOONEProvider({ mainRPCURL, testRPCURL, options });
  });

  test('NEOONEProvider constructor with no options', async () => {
    const testProvider = new NEOONEProvider();
    const result = await testProvider.getNetworks();
    expect(result).toEqual(['main', 'test']);
  });

  test('NEOONEProvider constructor with main & test', async () => {
    const testOptions = [
      { network: networkConfigs.MAIN, rpcURL: networkConfigs.MAIN_URL },
      { network: networkConfigs.TEST, rpcURL: networkConfigs.TEST_URL },
    ];

    const testExpected = testOptions.map((option) => option.network);

    // @ts-ignore
    const testProvider = new NEOONEProvider({ options: testOptions });
    const result = await testProvider.getNetworks();
    expect(result).toEqual(testExpected);
  });

  test('getNetworks', async () => {
    // @ts-ignore
    provider.networks$.getValue = jest.fn(() => '0');

    const result = provider.getNetworks();
    expect(result).toEqual(expected);
  });

  test('addNetwork new network', async () => {
    const newNetwork = 'newNet';
    provider.addNetwork({ network: newNetwork, rpcURL: 'rpc' });

    const result = await provider.getNetworks();
    expect(result).toEqual([network, 'main', 'test', newNetwork]);
  });

  test('addNetwork existing network', async () => {
    provider.addNetwork({ network, rpcURL: 'rpc' });

    const result = await provider.getNetworks();
    expect(result).toEqual(['main', 'test', network]);
  });

  test('read', () => {
    const result = provider.read(network);
    expect(result).toMatchSnapshot();
  });

  test('_getProvider throws error for unknown network', () => {
    const fakeNet = 'fake';
    function testError() {
      // @ts-ignore
      provider.getProvider(fakeNet);
    }
    expect(testError).toThrow(new UnknownNetworkError(fakeNet) as any);
  });

  const testCases = [
    {
      method: 'getUnclaimed',
      args: [network, keys[0].address],
    },

    {
      method: 'getUnspentOutputs',
      args: [network, keys[0].address],
    },

    {
      method: 'relayTransaction',
      args: [network, ''],
    },

    {
      method: 'getTransactionReceipt',
      args: [network, transactions.register.hash],
    },

    {
      method: 'getInvocationData',
      args: [network, transactions.register.hash],
    },

    {
      method: 'testInvoke',
      args: [network, ''],
    },

    {
      method: 'getNetworkSettings',
      args: [network],
    },
  ];

  for (const testCase of testCases) {
    const { method, args } = testCase;
    // eslint-disable-next-line
    test(method, async () => {
      // @ts-ignore
      provider.mutableProviders[network][method] = jest.fn(() => expected);
      // @ts-ignore
      const result = await provider[method](...args);
      expect(result).toEqual(expected);
    });
  }
});
