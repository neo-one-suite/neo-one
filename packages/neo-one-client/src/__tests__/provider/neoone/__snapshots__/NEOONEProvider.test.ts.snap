// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NEOONEProvider read 1`] = `
NEOONEDataProvider {
  "iterBlocksBatchSize": undefined,
  "iterBlocksFetchTimeoutMS": undefined,
  "mutableClient": JSONRPCClient {
    "provider": JSONRPCHTTPProvider {
      "batcher": DataLoader {
        "_batchLoadFn": [Function],
        "_options": Object {
          "cacheKeyFn": [Function],
          "maxBatchSize": 25,
        },
        "_promiseCache": Array [],
        "_queue": Array [],
      },
      "endpoint": "rpc",
    },
  },
  "network": "net",
}
`;
